File: Manejo de Excepciones/ejercicio 2/Inventario.py
class Inventario:
    def __init__(self):
        self.productos = []

    def agregarProducto(self, p):
        for prod in self.productos:
            if prod.codigo.lower() == p.codigo.lower():
                raise Exception(f"El código ya existe: {p.codigo}")

        if p.precio < 0 or p.stock < 0:
            raise Exception("Precio/stock no pueden ser negativos")

        self.productos.append(p)

    def buscarProducto(self, codigo):
        for p in self.productos:
            if p.codigo.lower() == codigo.lower():
                return p
        raise ProductoNoEncontradoException(f"Producto no encontrado: {codigo}")

    def venderProducto(self, codigo, cantidad):
        p = self.buscarProducto(codigo)
        if p.stock < cantidad:
            raise StockInsuficienteException("Stock insuficiente para vender " + str(cantidad))
        p.stock -= cantidad

    def listar(self):
        print("\nInventario:")
        for p in self.productos: 
            p.mostrar()
if __name__ == "__main__":
    inv = Inventario()
    try:
        inv.agregarProducto(Producto("P1", "Laptop", 2500, 5))
        inv.agregarProducto(Producto("P2", "Mouse", 40, 20))

        inv.listar()

        inv.venderProducto("P1", 3)
        inv.listar()

        inv.venderProducto("P1", 5)

    except Exception as e:
        print("Error:", e)

File: Manejo de Excepciones/ejercicio 2/Inventario.java
package manejo.de.excepciones;

import java.util.ArrayList;

public class Inventario {
    private ArrayList<Producto> productos = new ArrayList<>();

    public void agregarProducto(Producto p) throws Exception {
        for (Producto prod : productos) {
            if (prod.getCodigo().equalsIgnoreCase(p.getCodigo())) {
                throw new Exception("El código ya existe: " + p.getCodigo());
            }
        }
        if (p.getPrecio() < 0 || p.getStock() < 0) {
            throw new Exception("Precio/stock no pueden ser negativos");
        }
        productos.add(p);
    }

    public Producto buscarProducto(String codigo) throws ProductoNoEncontradoException{
        for (Producto p : productos)
            if (p.getCodigo().equalsIgnoreCase(codigo))
                return p;
        throw new ProductoNoEncontradoException("Producto no encontrado: " + codigo);
    }

    public void venderProducto(String codigo, int cantidad)
            throws ProductoNoEncontradoException, StockInsuficienteException {
        Producto p = buscarProducto(codigo);
        if (p.getStock() < cantidad)
            throw new StockInsuficienteException("Stock insuficiente para vender " + cantidad);

        p.setStock(p.getStock() - cantidad);
    }

    public void listar() {
        System.out.println("\nInventario:");
        for (Producto p : productos) p.mostrar();
    }
}